Test that bluechictl monitor receives UnitCreated/UnitRemoved signals

mwperina opened the pull request:
Adds a test, which verifies that bluechictl monitor is able to receive  UnitCreated and UnitRemoved signals.

Fixes: https://github.com/eclipse-bluechi/bluechi/issues/784
Signed-off-by: Martin Perina <mperina@redhat.com>

mwperina committed:
Test that bluechictl monitor receives UnitCreated/UnitRemoved signals

Adds a test, which verifies that bluechictl monitor is able to receive
UnitCreated and UnitRemoved signals.

Fixes: https://github.com/eclipse-bluechi/bluechi/issues/784
Signed-off-by: Martin Perina <mperina@redhat.com>

engelmi commented on tests/tests/tier0/bluechi-monitor-service-removed/test_bluechi_monitor_service_removed.py:
```diff
@@ -0,0 +1,56 @@
+
+    node_foo.systemctl.stop_unit(service_simple)
+    node_foo.exec_run(f"rm -f /etc/systemd/system/{service_simple}")
+    node_foo.systemctl.daemon_reload()
```
The `UnitRemoved` signal is emitted when a unit is unloaded (see Signals section [here](https://www.freedesktop.org/wiki/Software/systemd/dbus/)). So removing the file and daemon-reload aren't necessary.

mwperina replied:
Hmm, so I totally missed that and my interpretation of the event is wrong. So how can I invoke unit removed signal? Because according to my test it was invoked when I removed the file on the disk:

Service cups is running when monitoring is started:

```
[root@mperina-cs9 ~]# bluechictl monitor mperina-cs9-2 cups.service
Monitor path: /org/eclipse/bluechi/monitor/1
Subscribing to node 'mperina-cs9-2' and unit 'cups.service'
[mperina-cs9-2] cups.service
	Unit created (reason: virtual)
[mperina-cs9-2] cups.service
	Unit state changed (reason: virtual)
	Active: active (running)
[mperina-cs9-2] cups.service
	Unit properties changed (Interface: org.freedesktop.systemd1.Service)
	StatusText: Scheduler is running...
[mperina-cs9-2] cups.service
	Unit properties changed (Interface: org.freedesktop.systemd1.Unit)
	ActiveState: active
```

On package uninstallation the unit is stopped first:

```
[mperina-cs9-2] cups.service
	Unit properties changed (Interface: org.freedesktop.systemd1.Service)
	StatusText: Scheduler is running...
[mperina-cs9-2] cups.service
	Unit properties changed (Interface: org.freedesktop.systemd1.Unit)
	ActiveState: active
[mperina-cs9-2] cups.service
	Unit properties changed (Interface: org.freedesktop.systemd1.Service)
	StatusText: Scheduler is running...
[mperina-cs9-2] cups.service
	Unit state changed (reason: real)
	Active: deactivating (stop-sigterm)
[mperina-cs9-2] cups.service
	Unit properties changed (Interface: org.freedesktop.systemd1.Unit)
	ActiveState: deactivating
[mperina-cs9-2] cups.service
	Unit properties changed (Interface: org.freedesktop.systemd1.Service)
	StatusText: Scheduler is running...
[mperina-cs9-2] cups.service
	Unit state changed (reason: real)
	Active: inactive (dead)
[mperina-cs9-2] cups.service
	Unit properties changed (Interface: org.freedesktop.systemd1.Unit)
	ActiveState: inactive
```

And the service file is removed from disk:

```
[mperina-cs9-2] cups.service
	Unit properties changed (Interface: org.freedesktop.systemd1.Service)
	StatusText: Scheduler is running...
[mperina-cs9-2] cups.service
	Unit properties changed (Interface: org.freedesktop.systemd1.Unit)
	ActiveState: inactive
[mperina-cs9-2] cups.service
	Unit removed (reason: real)
[mperina-cs9-2] cups.service
	Unit created (reason: real)
[mperina-cs9-2] cups.service
	Unit removed (reason: real)
[mperina-cs9-2] cups.service
	Unit created (reason: real)
[mperina-cs9-2] cups.service
	Unit removed (reason: real)
```

So `Unit removed (reson: real)` was received only when the service file was removed from disk.

Do you know any other option how to raise UnitRemoved event?

mwperina replied:
UnitCreated and UnitRemoved signals are also received when `systemctl status` is invoked on `inactive` service, adapting the test accordingly

engelmi commented on tests/tests/tier0/bluechi-monitor-service-removed/main.fmf:
```diff
@@ -0,0 +1,2 @@
+summary: Test if a removal of a service can be monitored using bluechictl monitor
```
Its probably best to change the directory and test name to `bluechictl-monitor-service-removed` or something similar that includes the hint of using `bluechictl` since it could be (and is) done via the plain DBus API.

mwperina replied:
Yeah, I was thinking about it, so let's rename this one and in different PR rename other existing tests to cleary identify bluechictl usage

engelmi commented on tests/tests/tier0/bluechi-monitor-service-removed/test_bluechi_monitor_service_removed.py:
```diff
@@ -0,0 +1,56 @@
+    ctrl.bluechictl.start_unit(node_name_foo, service_simple)
+    nodes[node_name_foo].wait_for_unit_state_to_be(service_simple, "active")
+
+    t = threading.Thread(target=monitor_service, args=(ctrl,))
```
Is there a specific reason to start the monitoring script in a thread? Just running the script here and adding the short wait and unit stop in the monitor script (after starting the threads) seems a bit simpler.

mwperina replied:
But `run_python` is a blocking execution, right? If so, then I need to run it in a thread so I can remove the service when monitoring is running

engelmi replied:
Yes, it is a blocking call. 
You could move the `assert node_foo.systemctl.get_unit_state(service_simple) == "inactive"` in the `monitor.py` on line 107 (after starting the threads - that should work, I think, and you can remove the threading code from the test py. 
It is just a NIT, though.

mwperina replied:
Fixed

engelmi reviewed (approved):
LGTM

mwperina committed:
Use stdout for bluechictl status output

Information printed by bluechictl status should go to stdout, only
errors should go to stderr.

Signed-off-by: Martin Perina <mperina@redhat.com>
